kroxylicious/src/main/java/io/kroxylicious/proxy/internal/filter/EagerMetadataLearner.java

```diff
@@ -0,0 +1,81 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.kroxylicious.proxy.filter.KrpcFilterContext;
+import io.kroxylicious.proxy.filter.RequestFilter;
+
+/**
+ * An internal filter that causes the system to eagerly learn the cluster's topology by spontaneously emitting
+ * an out-of-band Metadata request at the earliest legal point in the Kafka conversation.  The response to allows
+ * the Endpoint reconciliation to take place so that restricted upstream bindings are replaced by true bindings to
+ * the actual upstream brokers.
+ * <br/>
+ * Once the bindings are made, the filter causes the client's connection to close.   This is done
+ * in order to force the client to reconnect, thus ensuring the client has a connection to the intended broker.
+ *
+ * @see io.kroxylicious.proxy.internal.net.EndpointRegistry
+ */
+public class EagerMetadataLearner implements RequestFilter {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(EagerMetadataLearner.class);
+
+    /**
+     * The set of the API keys that are permitted before the client would normally send a METADATA request.
+     */
+    private final static Set<ApiKeys> KAFKA_PRELUDE = Set.of(ApiKeys.API_VERSIONS, ApiKeys.SASL_HANDSHAKE, ApiKeys.SASL_AUTHENTICATE);
+
+    public EagerMetadataLearner() {
+    }
+
+    @Override
+    public void onRequest(ApiKeys apiKey, RequestHeaderData header, ApiMessage body, KrpcFilterContext filterContext) {
+        if (KAFKA_PRELUDE.contains(apiKey)) {
+            filterContext.forwardRequest(header, body);
+        }
+        else {
+            final short apiVersion = determineMetadataApiVersion(header);
+            // Send an out-of-band Metadata request. The response will be intercepted by the in-built BrokerAddressFilter.
+            // By the time control returns to the handler, the upstream addresses will have been reconciled.
+            var unused = filterContext.<MetadataResponseData> sendRequest(apiVersion, new MetadataRequestData())
+                    .thenAccept(metadataResponseData -> {
+                        if (apiKey.equals(ApiKeys.METADATA) && apiVersion == header.requestApiVersion()) {
+                            // The client's requested matched our out-of-band request, so we may as well return the
+                            // response.
+                            filterContext.forwardResponse(metadataResponseData);
+                        }
+                        // closing the connection is important. This client connection is connected to bootstrap (it could
+                        // be any broker or maybe not something else). we must close the connection to force the client to
+                        // connect again.
+                        filterContext.closeConnection();
+                        LOGGER.info("Closing upstream bootstrap connection {} now that endpoint reconciliation is complete.", filterContext.channelDescriptor());
```

Comment:
nit: reconciliation has been triggered we don't know its complete yet

Reply:
In general it does not make sense for filter B is start processing is filter A is still doing  processing that will determine its output.

In this specific, we do know reconciliation is complete, BrokerAddressFilter calls `join()` on the reconcile future. It blocks.

Thinking forward to once we have an async api, the same thing will apply.  BrokerAddressFilter won't resolve its response future until the reconcile is done. Kroxy won't invoke the next filter in the chain until it does.